### PR TITLE
[flang] remove support for std::complex value lowering.

### PR DESCRIPTION
**Pull request by jeanPerier**
Part of the [RFC to use mlir complex type](https://discourse.llvm.org/t/rfc-flang-replace-usages-of-fir-complex-by-mlir-complex-type/82292).

### CODE REVIEW

**Comment by DavidTruby** [comment]
I'm not familiar with the exact ABI of the `mlir::TupleType`, do they guarantee that it has the same ABI as a struct with the same members? I assume that is what it is underneath.
If so, LGTM.

Would this change allow us to use `std::complex` directly in interfaces in the runtime? That'd help us clean up the Windows stuff a lot (since `_Complex` doesn't exist there)

**Comment by DavidTruby** [comment]
On second thought, should we possibly remove the by-value models entirely? I think those will only cause issues. 
For an example off the top of my head, `std::complex<float>` and `struct {float; float;};` do not have the same return value ABI on Windows. So I think we should only ever be using `std::complex` by reference

**Comment by jeanPerier** [comment]
> I'm not familiar with the exact ABI of the mlir::TupleType, do they guarantee that it has the same ABI as a struct with the same members? I assume that is what it is underneath.

It lowers to `{ float, float }` in LLVM IR. I can't guarantee this is OK on all ABIs. MLIR is not really equipped to deal with aggregate by value ABIs. As you mentioned, we should rather stay away from this altogether.

> On second thought, should we possibly remove the by-value models entirely? I think those will only cause issues.

I agree with you here. It is currently only required for the REDUCE intrinsics, and more specifically for the callback signature. I think it is most likely a runtime bug since the callback is defined in Fortran, so I opened [an issue](https://github.com/llvm/llvm-project/issues/110674).

> For an example off the top of my head, `std::complex<float>` and `struct {float; float;};` do not have the same return value ABI on Windows.

I know this is true for _Complex vs `struct {float; float;};`, but the C++ standard defines std::complex as _class_ template that is storage compatible with `struct {float; float;}`. So if it is a class, it falls into the same ABI rules as a struct as far as I know. On 32 bit machine where _Complex and std::complex ABI differ, `std::complex` and `struct {float; float;};` are lowered the same way in intefaces (see [this godbolt example](https://godbolt.org/z/fvfsjb6ne)). Anyway, I agree with your point that it is best to stay away from it.

**Comment by jeanPerier** [comment]
@DavidTruby, I went with your suggestion and just removed the std::complex value support all together. The REDUCE implementation for complex functions needs to be fixed independently.

File: flang/include/flang/Optimizer/Builder/Runtime/RTBuilder.h [hunk 1]
```diff
@@ -400,17 +400,19 @@ constexpr TypeBuilderFunc getModel<bool &>() {
...
-constexpr TypeBuilderFunc getModel<std::complex<float>>() {
-  return [](mlir::MLIRContext *context) -> mlir::Type {
-    return mlir::ComplexType::get(mlir::FloatType::getF32(context));
-  };
-}
+
+// getModel<std::complex<T>> are not implemented on purpose.
+// Prefer passing/returning the complex by reference in the runtime to
+// avoid ABI issues.
+// C++ std::complex is not an intrinsic type, and it while it is storage
```
Comment (clementval) [inline flang/include/flang/Optimizer/Builder/Runtime/RTBuilder.h:407]:
```suggestion
// C++ std::complex is not an intrinsic type, and while it is storage
```

**Review by clementval — APPROVED** [review]
LGTM

**Pull request merged**